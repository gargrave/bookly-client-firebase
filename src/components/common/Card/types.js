// @flow
export type CardTextListItem = {
  title?: string,
  value?: string,
}
