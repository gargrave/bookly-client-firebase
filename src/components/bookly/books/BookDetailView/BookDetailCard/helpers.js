// @flow
import { format } from 'date-fns';

import type { Book } from '../../../../../globals/flowtypes';

const DATE_ONLY_FORMAT = 'MMM. DD, YYYY';
const DATE_AND_TIME_FORMAT = `${DATE_ONLY_FORMAT}, HH:mm:ss`;

export const bookBasicDates = (book: Book) => ([
  {
    title: 'Added',
    value: format(book.created, DATE_AND_TIME_FORMAT),
  },
  {
    title: 'Updated',
    value: format(book.updated, DATE_AND_TIME_FORMAT),
  },
]);

export const bookOptionalDates = (book: Book) => {
  const dates = [];
  if (book.startedOn) {
    dates.push({
      title: 'Started on',
      value: format(book.startedOn, DATE_ONLY_FORMAT),
    });
  }
  if (book.finishedOn) {
    dates.push({
      title: 'Finished on',
      value: format(book.finishedOn, DATE_ONLY_FORMAT),
    });
  }
  return dates;
};

export const hasOptionalDates = (book: Book) =>
  !!book.startedOn || !!book.finishedOn;
