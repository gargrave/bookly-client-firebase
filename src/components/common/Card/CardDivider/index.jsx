// @flow
import React from 'react';

import { buildClasses } from '../../../../utils/cssHelpers';

type Props = {
};

const CardDivider = ({}: Props) => {
  return (
    <hr className={buildClasses(['card__divider'])} />
  );
};

CardDivider.propTypes = {
};

export default CardDivider;
