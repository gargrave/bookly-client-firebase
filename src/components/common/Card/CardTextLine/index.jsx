// @flow
import React from 'react';
import { bool, string } from 'prop-types';

import { buildClasses } from '../../../../utils/cssHelpers';

type Props = {
  bold?: boolean,
  condensed?: boolean,
  text: string,
};

const CardTextLine = ({
  bold = false,
  condensed = false,
  text,
}: Props) => {
  return (
    <p
      className={buildClasses([
        'card__text',
        bold ? 'card__text--bold' : '',
        condensed ? 'condensed' : '',
      ])}
    >
      {text}
    </p>
  );
};

CardTextLine.propTypes = {
  bold: bool,
  condensed: bool,
  text: string.isRequired,
};

export default CardTextLine;
