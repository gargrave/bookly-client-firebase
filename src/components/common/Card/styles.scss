@import '../../../styles/_index';

.bookly-card {
  @include boxShadowMed;
  background-color: #f9f9f9;
  border: solid 1px #bbb;
  border-radius: 5px;
  margin: 8px auto;
  max-width: 600px;
  padding: 10px 20px;

  &--top-margin-med {
    margin-top: 50px;
  }

  hr {
    margin: 12px 0;
  }

  &--hoverable:hover {
    background: #e4effc;
    cursor: pointer;
  }

  &__header,
  &__title,
  &__text {
    margin: 0;
  }

  &__header,
  &__title {
    font-weight: bold;
  }

  &__header {
    font-size: 1.5rem;
  }

  &__text {
    color: #6e6969;
  }

  &__text + &__text {
    margin-top: 8px;
  }
}

.bookly-detail-card {
  padding: 15px;

  .bookly-card__title {
    font-size: 1.3rem;
    line-height: 2.5rem;
  }

  .bookly-card__title + .bookly-card__text {
    font-size: 1.1rem;
  }
}

.bookly {
  &-list-detail-card {
    padding: 10px
  }
}
