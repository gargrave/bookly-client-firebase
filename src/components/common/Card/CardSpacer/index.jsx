// @flow
import React from 'react';
import { oneOf } from 'prop-types';

import { buildClasses } from '../../../../utils/cssHelpers';

import './styles.css';

type Props = {
  size?: string,
};

const CardSpacer = ({
  size = 'medium',
}: Props) => {
  return (
    <div className={buildClasses(`card__spacer--${size}`)} />
  );
};

CardSpacer.propTypes = {
  size: oneOf(['small', 'medium', 'large']),
};

export default CardSpacer;
