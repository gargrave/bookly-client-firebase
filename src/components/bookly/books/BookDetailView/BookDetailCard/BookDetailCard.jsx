// @flow
import React, { Fragment } from 'react';
import { func, object } from 'prop-types';

import type { Book } from '../../../../../globals/flowtypes';

import AuthorLink from '../../../authors/AuthorLink';
import Button from '../../../../common/Button';
import ButtonRow from '../../../../common/ButtonRow';
import Card from '../../../../common/Card/Card';
import CardSpacer from '../../../../common/Card/CardSpacer/CardSpacer';
import CardTextList from '../../../../common/Card/CardTextList/CardTextList';

import {
  bookBasicDates,
  bookOptionalDates,
  hasOptionalDates,
} from './helpers';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      classes={['detail-card', 'book-detail-card']}
      header={title}
      hoverable={false}
    >
      <AuthorLink author={book.author} />

      {hasOptionalDates(book) &&
        <Fragment>
          <CardTextList textList={bookOptionalDates(book)} />
          <CardSpacer />
        </Fragment>
      }
      <CardTextList textList={bookBasicDates(book)} />
      <CardSpacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info"
        />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger"
        />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light"
        />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: object.isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
