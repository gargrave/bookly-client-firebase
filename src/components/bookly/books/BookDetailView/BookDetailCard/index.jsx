// @flow
import React from 'react';
import { func, instanceOf, oneOfType, shape, string } from 'prop-types';
import { format } from 'date-fns';

import type { Book } from '../../../../../constants/flowtypes';

import AuthorLink from '../../../authors/AuthorLink';
import Button from '../../../../common/Button';
import ButtonRow from '../../../../common/ButtonRow';
import Card from '../../../../common/Card';
import CardSpacer from '../../../../common/Card/CardSpacer';
import CardTextList from '../../../../common/Card/CardTextList';

type Props = {
  book: Book,
  onBackClick: Function,
  onDeleteClick: Function,
  onEditClick: Function,
};

const bookDatesTextList = (book: Book) => {
  const {
    created,
    updated,
  } = book;

  return [
    { title: 'Added', value: format(created, 'MMM. DD, YYYY, HH:mm:ss') },
    { title: 'Updated', value: format(updated, 'MMM. DD, YYYY, HH:mm:ss') },
  ];
};

const BookDetailCard = ({
  book,
  onBackClick,
  onDeleteClick,
  onEditClick,
}: Props) => {
  const {
    title,
  } = book;

  return (
    <Card
      classes={['detail-card', 'book-detail-card']}
      header={title}
      hoverable={false}
    >
      <AuthorLink
        author={book.author}
      />

      <CardTextList textList={bookDatesTextList(book)} />
      <CardSpacer size='large' />

      <ButtonRow>
        <Button
          onClick={onEditClick}
          position="left"
          text="Edit"
          type="info"
        />
        <Button
          onClick={onDeleteClick}
          text="Delete"
          type="danger"
        />
        <Button
          onClick={onBackClick}
          position="right"
          text="Back"
          type="light"
        />
      </ButtonRow>
    </Card>
  );
};

BookDetailCard.propTypes = {
  book: shape({
    created: oneOfType([
      instanceOf(Date),
      string,
    ]),
    updated: oneOfType([
      instanceOf(Date),
      string,
    ]),
    author: shape({
      firstName: string,
      lastName: string,
    }),
    title: string,
  }).isRequired,
  onBackClick: func.isRequired,
  onDeleteClick: func.isRequired,
  onEditClick: func.isRequired,
};

export default BookDetailCard;
