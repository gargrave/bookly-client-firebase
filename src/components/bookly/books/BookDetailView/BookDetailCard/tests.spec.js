import React from 'react';
import { shallow } from 'enzyme';

import { bookMocks } from '../../../../../utils/mocks/';

import Card from '../../../../common/Card';

import BookDetailCard from './';

const defaultProps = Object.freeze({
  book: Object.create(bookMocks[0]),
  onBackClick: jest.fn(),
  onDeleteClick: jest.fn(),
  onEditClick: jest.fn(),
});

function getComponent(extraProps = {}) {
  const props = Object.assign({}, defaultProps, extraProps);
  return shallow(<BookDetailCard {...props} />);
}

describe('BookDetailCard', () => {
  let component;

  test('matches the snapshot', () => {
    component = getComponent();
    expect(component).toMatchSnapshot();
  });

  test('renders correctly', () => {
    component = getComponent();
    expect(component.find(Card)).toHaveLength(1);
  });
});
