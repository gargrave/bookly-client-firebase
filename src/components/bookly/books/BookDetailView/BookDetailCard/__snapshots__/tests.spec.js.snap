// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetailCard matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookDetailCard
    book={Object {}}
    onBackClick={[Function]}
    onDeleteClick={[Function]}
    onEditClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <AuthorLink
          author={
            Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            }
          }
        />,
        <hr />,
        <p
          className="bookly-card-text"
        >
          <strong>
            Added:
          </strong>
           
          Jan. 13, 2018, 17:36:26
        </p>,
        <p
          className="bookly-card-text"
        >
          <strong>
            Updated:
          </strong>
           
          Jan. 13, 2018, 17:36:26
        </p>,
        <hr />,
        <ButtonRow>
          <Button
            onClick={[Function]}
            position="left"
            text="Edit"
            type="info"
          />
          <Button
            onClick={[Function]}
            text="Delete"
            type="danger"
          />
          <Button
            onClick={[Function]}
            position="right"
            text="Back"
            type="light"
          />
        </ButtonRow>,
      ],
      "classes": Array [
        "card--top-margin-med",
        "detail-card",
        "book-detail-card",
      ],
      "hoverable": false,
      "title": "Brave New World",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "bookCount": 1,
            "created": 2017-07-01T01:16:43.881Z,
            "firstName": "Aldous",
            "id": "oiw223h3qndlksadffl3ksdf",
            "lastName": "Huxley",
            "updated": 2017-07-01T01:16:43.881Z,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <strong>
              Added:
            </strong>,
            " ",
            "Jan. 13, 2018, 17:36:26",
          ],
          "className": "bookly-card-text",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Added:",
            },
            "ref": null,
            "rendered": "Added:",
            "type": "strong",
          },
          " ",
          "Jan. 13, 2018, 17:36:26",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <strong>
              Updated:
            </strong>,
            " ",
            "Jan. 13, 2018, 17:36:26",
          ],
          "className": "bookly-card-text",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Updated:",
            },
            "ref": null,
            "rendered": "Updated:",
            "type": "strong",
          },
          " ",
          "Jan. 13, 2018, 17:36:26",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Button
              onClick={[Function]}
              position="left"
              text="Edit"
              type="info"
            />,
            <Button
              onClick={[Function]}
              text="Delete"
              type="danger"
            />,
            <Button
              onClick={[Function]}
              position="right"
              text="Back"
              type="light"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "position": "left",
              "text": "Edit",
              "type": "info",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "text": "Delete",
              "type": "danger",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "position": "right",
              "text": "Back",
              "type": "light",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <AuthorLink
            author={
              Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              }
            }
          />,
          <hr />,
          <p
            className="bookly-card-text"
          >
            <strong>
              Added:
            </strong>
             
            Jan. 13, 2018, 17:36:26
          </p>,
          <p
            className="bookly-card-text"
          >
            <strong>
              Updated:
            </strong>
             
            Jan. 13, 2018, 17:36:26
          </p>,
          <hr />,
          <ButtonRow>
            <Button
              onClick={[Function]}
              position="left"
              text="Edit"
              type="info"
            />
            <Button
              onClick={[Function]}
              text="Delete"
              type="danger"
            />
            <Button
              onClick={[Function]}
              position="right"
              text="Back"
              type="light"
            />
          </ButtonRow>,
        ],
        "classes": Array [
          "card--top-margin-med",
          "detail-card",
          "book-detail-card",
        ],
        "hoverable": false,
        "title": "Brave New World",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "bookCount": 1,
              "created": 2017-07-01T01:16:43.881Z,
              "firstName": "Aldous",
              "id": "oiw223h3qndlksadffl3ksdf",
              "lastName": "Huxley",
              "updated": 2017-07-01T01:16:43.881Z,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <strong>
                Added:
              </strong>,
              " ",
              "Jan. 13, 2018, 17:36:26",
            ],
            "className": "bookly-card-text",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Added:",
              },
              "ref": null,
              "rendered": "Added:",
              "type": "strong",
            },
            " ",
            "Jan. 13, 2018, 17:36:26",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <strong>
                Updated:
              </strong>,
              " ",
              "Jan. 13, 2018, 17:36:26",
            ],
            "className": "bookly-card-text",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Updated:",
              },
              "ref": null,
              "rendered": "Updated:",
              "type": "strong",
            },
            " ",
            "Jan. 13, 2018, 17:36:26",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Button
                onClick={[Function]}
                position="left"
                text="Edit"
                type="info"
              />,
              <Button
                onClick={[Function]}
                text="Delete"
                type="danger"
              />,
              <Button
                onClick={[Function]}
                position="right"
                text="Back"
                type="light"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "position": "left",
                "text": "Edit",
                "type": "info",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "text": "Delete",
                "type": "danger",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "position": "right",
                "text": "Back",
                "type": "light",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
